language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashSet;
+use std::convert::TryFrom;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use anyhow::{bail, format_err, Result};
+use diem_client::{
+    views::AccountStateWithProofView, AccountAddress, BlockingClient, MethodRequest, MethodResponse,
+};
+use diem_read_write_set::ReadWriteSetAnalysis;
+use diem_state_view::StateView;
+use diem_transaction_replay::DiemDebugger;
+use diem_types::access_path::AccessPath;
+use diem_types::transaction::{TransactionOutput, Version};
+use diem_types::{
+    account_state::AccountState, account_state_blob::AccountStateBlob, transaction::Transaction,
+};
+use diem_validator_interface::{DiemValidatorInterface, JsonRpcDebuggerInterface};
+use diem_vm::VMExecutor;
+use move_cli::sandbox::utils::{
+    mode::{Mode, ModeType},
+    on_disk_state_view::OnDiskStateView,
+};
+use move_core_types::language_storage::ResourceKey;
+
+use structopt::StructOpt;
+
+pub struct OnDiskStateViewReader {
+    view: OnDiskStateView,
+}
+
+struct AccountAddressWithState {
+    address: AccountAddress,
+    state: AccountState,
+}
+
+#[derive(Debug)]
+pub struct PrefetchTransactionReplayerConfig {
+    build_dir: PathBuf,
+    storage_dir: PathBuf,
+    rpc_url: String,
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    about = "Replay transactions using state prefetching",
+    rename_all = "kebab-case"
+)]
+pub struct PrefetchedReplayCLI {
+    #[structopt(long, default_value = "http://localhost:8080")]
+    rpc_url: String,
+
+    #[structopt(subcommand)]
+    cmd: Command,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum Command {
+    /// Replay transactions starting from version `start` to `start + limit`.
+    #[structopt(name = "replay-transactions")]
+    ReplayTransactions { start: Version, limit: u64 },
+}
+
+pub struct PrefetchTransactionReplayer {
+    config: PrefetchTransactionReplayerConfig,
+    validator: Box<dyn DiemValidatorInterface>,
+    debugger: DiemDebugger,
+    rpc_client: BlockingClient,
+}
+
+impl StateView for OnDiskStateViewReader {
```

Comment:
I think it would be preferable to implement this trait on `OnDiskStateView`--you won't need to do an extra serialization/deserialization roundtrip that way.

Reply:
Daniel correctly noted that `StateView` is Diem-specific, so the Move CLI (where `OnDiskStateView` lives) can't implement it.